util/dtgen/dt_api.h.tpl

```diff
@@ -0,0 +1,377 @@
+
+/** PLIC IRQ ID for no interrupt. */
+static const dt_plic_irq_id_t kDtPlicIrqIdNone=${top_name.as_c_enum()}PlicIrqIdNone;
+
+/**
+ * Get the instance ID for a given PLIC IRQ ID.
+ *
+ * For example, on earlgrey, the instance ID of `kTopEarlgreyPlicIrqIdUart0TxWatermark`
+ * is `kDtInstanceIdUart0`. One can then use the type specific function to retrieve the
+ * IRQ name, for example `dt_uart_irq_from_plic_id` for the UART.
+ *
+ * @param dev A PLIC ID.
+ * @return The instance ID, or `kDtInstanceIdUnknown` if the PLIC ID is not valid.
+ */
+dt_instance_id_t dt_plic_id_to_instance_id(dt_plic_irq_id_t irq);
+
+/**
+ * List of clocks.
+ *
+ * Clocks are guaranteed to be numbered consecutively from 0.
+ */
+${helper.clock_enum.render()}
+
+/**
+ * Get the frequency of a clock.
+ *
+ * @param dev A clock ID.
+ * @return Clock frequency in Hz.
+ */
+uint32_t dt_clock_frequency(dt_clock_t clk);
+
+/**
+ * Pinmux types.
+ *
+ * These types are aliases to top-level types for backward compatibility
+ * with existing code.
+ */
+typedef ${top_name.as_snake_case()}_pinmux_peripheral_in_t dt_pinmux_peripheral_in_t;
+typedef ${top_name.as_snake_case()}_pinmux_insel_t dt_pinmux_insel_t;
+typedef ${top_name.as_snake_case()}_pinmux_outsel_t dt_pinmux_outsel_t;
+typedef ${top_name.as_snake_case()}_pinmux_mio_out_t dt_pinmux_mio_out_t;
+typedef ${top_name.as_snake_case()}_direct_pads_t dt_pinmux_direct_pad_t;
+typedef ${top_name.as_snake_case()}_muxed_pads_t  dt_pinmux_muxed_pad_t;
+
+/** Type of a pin. */
+typedef enum dt_pin_type {
+  /* This pin is connected to a muxed IO (MIO). */
+  kDtPinTypeMio,
+  /* This pin is connected to a direct IO (DIO). */
+  kDtPinTypeDio,
+  /* This pin is not connected to either a MIO or a DIO. */
+  kDtPinTypeOther,
+} dt_pin_type_t;
+
+/** Pin description.
+ *
+ * A `dt_pin_t` represents a HW IP block pin, which can be an input, output or both.
+ * Importantly, this only represents how the block pin is wired to the pinmux, i.e.
+ * whether it is a MIO or a direct IO, and the relevant information necessary to
+ * configure it.
+ *
+ * NOTE The fields of this structure are internal, use the dt_pin_* functions to access them.
+ */
+typedef struct dt_pin {
+  struct {
+    /** Pin type */
+    uint16_t type : 2;
+    /** For `kDtPinTypeMio` pins: peripheral input number. This is the index of the MIO_PERIPH_INSEL register
+     * that controls this pin (or the input part of this pin). Set to `kDtPinPeriphInputNone`
+     * if this pin is not an input.
+     *
+     * For `kDtPinTypeDio`:  DIO pad number. This is the index of the various DIO_PAD_* registers
+     * that control this pin.
+     */
+    uint16_t periph_input_or_direct_pad: 7;
+    /** For `kDtPinTypeMio` pins: peripheral output number. This is the value to put in the MIO_OUTSEL registers
+     * to connect an output to this pin. Set to `kDtPinOutselNone` if this pin is not an input.
+     */
+    uint16_t outsel: 7;
```

Comment:
Just a note that pins or "signals" are generally unidirectional, so this would probably confuse more than help in checked in files.

But it's not checked in, so... /shrug